actions/shell_action.go

```diff
@@ -0,0 +1,100 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package actions
+
+import (
+	"context"
+	"os/exec"
+	"sync"
+)
+
+type shellAction struct {
+	id      string
+	command string
+	arg     []string
+
+	forbidden map[string]struct{}
+
+	// Because cxt, and cancel are using in Run() which is blocked and Stop(),
+	// and because those methods can be used by separate goroutine
+	// we should protect those vars by Mutex.
+	mx     sync.Mutex
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewShellAction creates Shell action.
+//
+// Shell action, it's an abstract action that can run some "predefined" set of shell commands.
+// This commands can be a shell script, script written on interpreted language, or binary file.
+func NewShellAction(id string, command string, arg []string) Action {
+	return &shellAction{
+		id:      id,
+		command: command,
+		arg:     arg,
+		forbidden: map[string]struct{}{
```

Comment:
For security reasons, we can't have this in place. we need an only list(mapping) from command to bash script. No  direct name matching

Reply:
@rnovikovP Yes but this binary path and name are configurable (They can be set with `--paths-...` flags), So we don't know how it will be called. 

Look how it used, here we already have that mapping you want:

https://github.com/percona/pmm-agent/pull/47/files#diff-1cd41a2611da49ce63630d5a51bd82c5R262

Reply:
It seems this check is redundant at all as we check action type at a different level. Also, I use some commands like `sleep` and `echo` for testing. We can leave it as-is, remove it, or I can do something like this:

```golang
return &shellAction{
		id:      id,
		command: command,
		arg:     arg,
		allowed: map[string]struct{}{
			"pt-summary":       {},
			"pt-mysql-summary": {},
			"sleep":            {}, // for testing
			"echo":             {}, // for testing
		},
	}

	_, file := path.Split(p.command)
	if _, ok := p.allowed[file]; !ok {
		return nil, errUnknownAction
	}
```
But then we should change this internal library implementation all the time we add a new action, and I think it's a bad thing. Also, it breaks the possibility to give custom names to shell commands actions.

Reply:
1. This type of verification is confusing. It  looks like  you  can pass anything and some developer later can expect this  
1. Negative verification if what we need to use only in rare cases. This is very dangerous
1. If we don't need it at all ...maybe better to remove 
1. If there were good to have some Action without parameters and not dependent on binary installed on the Client, i would suggest having something like `uname` command.  It will be useful later ;)